### PR TITLE
Refactor reserves

### PR DESCRIPTION
**Pull request by Nnamdichukwu**
1. Refactored the functions(Deposit, Withdraw, Repay, Borrow, ExecuteBorrow) LendingPool.sol to accept an additional pool param that is equal to the lending pool addresss.
2. Refactored Reserves and UserConfig in LendingPoolStorage.sol
3. Refactored LendingPoolAddressesProvider to set PoolAdmin, Pool Address dynamically.
4. Added Lending_Pool to the constructor of LendingPoolAddressesProvider.sol
5. Refactored setMarketId and getMarketId to accept bytes32 LendingPool

### CODE REVIEW

File: contracts/interfaces/ILendingPoolAddressesProvider.sol [hunk 1]
```diff
@@ -9,7 +9,7 @@ pragma solidity 0.6.12;
  * @author Aave
  **/
 interface ILendingPoolAddressesProvider {
-  event MarketIdSet(string newMarketId);
+  event MarketIdSet(string newMarketId, bytes32 newLendingPOOL);
```
Comment (akinmail) [inline contracts/interfaces/ILendingPoolAddressesProvider.sol]: Remove market event

File: contracts/protocol/configuration/LendingPoolAddressesProvider.sol [hunk 2]
```diff
@@ -20,16 +20,17 @@ contract LendingPoolAddressesProvider is Ownable, ILendingPoolAddressesProvider
...
   bytes32 private constant LENDING_POOL_COLLATERAL_MANAGER = 'COLLATERAL_MANAGER';
   bytes32 private constant PRICE_ORACLE = 'PRICE_ORACLE';
   bytes32 private constant LENDING_RATE_ORACLE = 'LENDING_RATE_ORACLE';
 
-  constructor(string memory marketId) public {
-    _setMarketId(marketId);
+  constructor(string memory marketId, bytes32 memory LENDING_POOL) public {
+     
+    _setMarketId(marketId, LENDING_POOL);
   }
```
Comment (akinmail) [inline contracts/protocol/configuration/LendingPoolAddressesProvider.sol]: Remove `market_id` but leave `LENDING_POOL`

File: contracts/protocol/configuration/LendingPoolAddressesProvider.sol [hunk 3]
```diff
@@ -20,16 +20,17 @@ contract LendingPoolAddressesProvider is Ownable, ILendingPoolAddressesProvider
   string private _marketId;
```
Comment (akinmail) [inline contracts/protocol/configuration/LendingPoolAddressesProvider.sol]: Remove `marketId`

File: contracts/protocol/configuration/LendingPoolAddressesProvider.sol [hunk 4]
```diff
@@ -39,13 +40,19 @@ contract LendingPoolAddressesProvider is Ownable, ILendingPoolAddressesProvider
...
   /**
    * @dev Allows to set the market which this LendingPoolAddressesProvider represents
    * @param marketId The market id
+   * @param LENDING_POOL  The Lending Pool 
    */
-  function setMarketId(string memory marketId) external override onlyOwner {
-    _setMarketId(marketId);
+  function setMarketId(string memory marketId, bytes32  LENDING_POOL) external override onlyOwner {
+    _setMarketId(marketId, LENDING_POOL);
   }
```
Comment (akinmail) [inline contracts/protocol/configuration/LendingPoolAddressesProvider.sol]: Remove all functions to get or set markets as market variable has been fully removed

File: contracts/protocol/configuration/LendingPoolAddressesProvider.sol [hunk 5]
```diff
@@ -208,8 +218,9 @@ contract LendingPoolAddressesProvider is Ownable, ILendingPoolAddressesProvider
...
 
-  function _setMarketId(string memory marketId) internal {
+  function _setMarketId(string memory marketId, bytes32 _LENDING_POOL) internal {
     _marketId = marketId;
-    emit MarketIdSet(marketId);
+    LENDING_POOL = _LENDING_POOL;
+    emit MarketIdSet(marketId, _LENDING_POOL);
   }
 }
```
Comment (akinmail) [inline contracts/protocol/configuration/LendingPoolAddressesProvider.sol]: Remove this as well

File: contracts/interfaces/ILendingPoolAddressesProvider.sol [hunk 6]
```diff
@@ -9,7 +9,7 @@ pragma solidity 0.6.12;
  * @author Aave
  **/
 interface ILendingPoolAddressesProvider {
-  event MarketIdSet(string newMarketId);
```
Comment (akinmail) [inline contracts/interfaces/ILendingPoolAddressesProvider.sol:12]: Anytime you modify any function, make sure to also modify the interface

**Review by akinmail — CHANGES_REQUESTED** [review]
Please see changes + see notion page additions https://www.notion.so/Blessing-Demand-User-Story-Breakdown-bcf9b06f913242078d098d44ce2f141e#460429ba835b4071bc10a5afd71ed40e

File: contracts/interfaces/ILendingPool.sol [hunk 7]
```diff
@@ -9,13 +9,15 @@ interface ILendingPool {
...
    * @param reserve The address of the underlying asset of the reserve
+   * @param pool The address of the underlying pool of the reserve  
    * @param user The address initiating the deposit
    * @param onBehalfOf The beneficiary of the deposit, receiving the aTokens
    * @param amount The amount deposited
    * @param referral The referral code used
    **/
   event Deposit(
-    address indexed reserve,
+    address  reserve,
```
Comment (akinmail) [inline contracts/interfaces/ILendingPool.sol]: Why did you remove indexed?

File: contracts/interfaces/ILendingPool.sol [hunk 8]
```diff
@@ -9,13 +9,15 @@ interface ILendingPool {
...
+    address index pool,
```
Comment (akinmail) [inline contracts/interfaces/ILendingPool.sol]: Make pool indexed as well

File: contracts/interfaces/ILendingPool.sol [hunk 9]
```diff
@@ -25,15 +27,17 @@ interface ILendingPool {
   /**
    * @dev Emitted on withdraw()
    * @param reserve The address of the underlyng asset being withdrawn
+   * @param pool The address of the underlying pool of the reserve 
    * @param user The address initiating the withdrawal, owner of aTokens
    * @param to Address that will receive the underlying
    * @param amount The amount to be withdrawn
    **/
-  event Withdraw(address indexed reserve, address indexed user, address indexed to, uint256 amount);
+  event Withdraw(address  reserve, address indexed pool, address indexed user, address indexed to, uint256 amount);
```
Comment (akinmail) [inline contracts/interfaces/ILendingPool.sol]: Why did you remove indexed?

File: contracts/interfaces/ILendingPool.sol [hunk 10]
```diff
@@ -43,7 +47,8 @@ interface ILendingPool {
    * @param referral The referral code used
    **/
   event Borrow(
-    address indexed reserve,
+    address  reserve,
```
Comment (akinmail) [inline contracts/interfaces/ILendingPool.sol]: Same as above

File: contracts/interfaces/ILendingPool.sol [hunk 11]
```diff
@@ -55,14 +60,16 @@ interface ILendingPool {
   /**
    * @dev Emitted on repay()
    * @param reserve The address of the underlying asset of the reserve
+   * @param pool The address of the underlying pool of the reserve 
    * @param user The beneficiary of the repayment, getting his debt reduced
    * @param repayer The address of the user initiating the repay(), providing the funds
    * @param amount The amount repaid
    **/
   event Repay(
-    address indexed reserve,
```
Comment (akinmail) [inline contracts/interfaces/ILendingPool.sol:63]: Same as above

File: contracts/interfaces/ILendingPool.sol [hunk 12]
```diff
@@ -9,13 +9,15 @@ interface ILendingPool {
...
    * @param amount The amount deposited
    * @param referral The referral code used
    **/
   event Deposit(
-    address indexed reserve,
+    address  reserve,
+    address index pool,
     address user,
     address indexed onBehalfOf,
     uint256 amount,
```
Comment (akinmail) [inline contracts/interfaces/ILendingPool.sol:23]: Sacrifice indexed on referral for indexed on pool

File: contracts/interfaces/ILendingPool.sol [hunk 13]
```diff
@@ -400,7 +420,7 @@ interface ILendingPool {
     uint256 balanceToBefore
   ) external;
 
-  function getReservesList() external view returns (address[] memory);
+  function getReservesList(address asset) external view returns (address[] memory);
```
Comment (akinmail) [inline contracts/interfaces/ILendingPool.sol]: Suspect?

File: contracts/interfaces/ILendingPool.sol [hunk 14]
```diff
@@ -353,9 +372,10 @@ interface ILendingPool {
   /**
    * @dev Returns the configuration of the reserve
    * @param asset The address of the underlying asset of the reserve
+   * @param pool The address of the pool of the reserve
    * @return The configuration of the reserve
    **/
-  function getConfiguration(address asset)
+  function getConfiguration(address pool, address asset)
```
Comment (akinmail) [inline contracts/interfaces/ILendingPool.sol:378]: Suspect?

File: contracts/protocol/lendingpool/LendingPool.sol [hunk 15]
```diff
@@ -689,11 +698,11 @@ contract LendingPool is VersionedInitializable, ILendingPool, LendingPoolStorage
   /**
    * @dev Returns the list of the initialized reserves
    **/
-  function getReservesList() external view override returns (address[] memory) {
+  function getReservesList(address asset) external view override returns (address[] memory) {
     address[] memory _activeReserves = new address[](_reservesCount);
 
     for (uint256 i = 0; i < _reservesCount; i++) {
-      _activeReserves[i] = _reservesList[i];
+      _activeReserves[i][asset] = _reservesList[i][asset];
```
Comment (akinmail) [inline contracts/protocol/lendingpool/LendingPool.sol]: Make `_activeReserves` 2 dimentional

File: contracts/protocol/lendingpool/LendingPoolConfigurator.sol [hunk 16]
```diff
@@ -123,7 +124,7 @@ contract LendingPoolConfigurator is VersionedInitializable, ILendingPoolConfigur
     );
 
     DataTypes.ReserveConfigurationMap memory currentConfig =
-      pool.getConfiguration(input.underlyingAsset);
+      pool.getConfiguration(input.poolAddress, input.underlyingAsset);
```
Comment (akinmail) [inline contracts/protocol/lendingpool/LendingPoolConfigurator.sol]: make it `input.pool`

File: contracts/misc/UiPoolDataProviderV2V3.sol [hunk 17]
```diff
@@ -72,7 +72,7 @@ contract UiPoolDataProviderV2V3 is IUiPoolDataProviderV3 {
   {
     IAaveOracle oracle = IAaveOracle(provider.getPriceOracle());
     ILendingPool lendingPool = ILendingPool(provider.getLendingPool());
-    address[] memory reserves = lendingPool.getReservesList();
+    address[] memory reserves = lendingPool.getReservesList(address asset);
     AggregatedReserveData[] memory reservesData = new AggregatedReserveData[](reserves.length);
 
     for (uint256 i = 0; i < reserves.length; i++) {
```
Comment (akinmail) [inline contracts/misc/UiPoolDataProviderV2V3.sol:80]: check where `getReservesList` is being called in this file and modify

**Review by akinmail — CHANGES_REQUESTED** [review]
3rd batch of changes

File: contracts/interfaces/ILendingPool.sol (hunk 7 above)
Comment (Nnamdichukwu) [reply contracts/interfaces/ILendingPool.sol]: Resolved. Made reserve indexed

File: contracts/interfaces/ILendingPool.sol (hunk 9 above)
Comment (Nnamdichukwu) [reply contracts/interfaces/ILendingPool.sol]: Same comment as above.

File: contracts/interfaces/ILendingPool.sol (hunk 7 above)
Comment (Nnamdichukwu) [reply contracts/interfaces/ILendingPool.sol]: This has been resolved. Sacrificed `user` to make reserve indexed

File: contracts/interfaces/ILendingPool.sol (hunk 9 above)
Comment (Nnamdichukwu) [reply contracts/interfaces/ILendingPool.sol]: this has been resolved.

File: contracts/interfaces/ILendingPool.sol (hunk 10 above)
Comment (Nnamdichukwu) [reply contracts/interfaces/ILendingPool.sol]: this has been resolved, sacrificed `user` to make reserve indexed

File: contracts/interfaces/ILendingPool.sol (hunk 11 above)
Comment (Nnamdichukwu) [reply contracts/interfaces/ILendingPool.sol:63]: resolved. sacrificed `pool` to make reserve indexed

File: contracts/protocol/lendingpool/LendingPool.sol (hunk 15 above)
Comment (Nnamdichukwu) [reply contracts/protocol/lendingpool/LendingPool.sol]: this has been fixed

File: contracts/protocol/lendingpool/LendingPoolConfigurator.sol (hunk 16 above)
Comment (Nnamdichukwu) [reply contracts/protocol/lendingpool/LendingPoolConfigurator.sol]: this has been fixed

File: contracts/misc/UiPoolDataProviderV2V3.sol (hunk 17 above)
Comment (Nnamdichukwu) [reply contracts/misc/UiPoolDataProviderV2V3.sol:80]: i modified this function to accept the pool address as this function is meant to return an address of active reserves(that is assets that are active in that pool address). this solution isn't the most ideal as i think it can be better especially in the LendingPool.sol implementation

File: contracts/interfaces/ILendingPool.sol (hunk 7 above)
Comment (akinmail) [reply contracts/interfaces/ILendingPool.sol]: No thats not what you did, you sacrifised indexed in refferal  for pool here. This is deposit event

File: contracts/interfaces/ILendingPool.sol (hunk 10 above)
Comment (akinmail) [reply contracts/interfaces/ILendingPool.sol]: No thats not what you did, you sacrifised indexed in refferal for pool here. This is Borrow event

File: contracts/interfaces/ILendingPool.sol (hunk 11 above)
Comment (akinmail) [reply contracts/interfaces/ILendingPool.sol:63]: No indexes was sacrifised here

File: contracts/misc/UiPoolDataProviderV2V3.sol [hunk 18]
```diff
@@ -72,7 +72,7 @@ contract UiPoolDataProviderV2V3 is IUiPoolDataProviderV3 {
   {
     IAaveOracle oracle = IAaveOracle(provider.getPriceOracle());
     ILendingPool lendingPool = ILendingPool(provider.getLendingPool());
-    address[] memory reserves = lendingPool.getReservesList(address asset);
+    address[] memory reserves = lendingPool.getReservesList(address pool);
```
Comment (akinmail) [inline contracts/misc/UiPoolDataProviderV2V3.sol]: Where is pool from? where is it forst defined or imported?

**Review by akinmail — CHANGES_REQUESTED** [review]
last changes

File: contracts/misc/UiPoolDataProviderV2V3.sol (hunk 18 above)
Comment (Nnamdichukwu) [reply contracts/misc/UiPoolDataProviderV2V3.sol]: it is `address pool`so it is a param that's created inside that function.

Comment (akinmail) [reply contracts/misc/UiPoolDataProviderV2V3.sol]: What line was it created? Cant find it

**Review by akinmail — CHANGES_REQUESTED** [review]
more questions

**Review by akinmail — APPROVED** [review]
LGTM
